paper-server/src/main/java/org/bukkit/craftbukkit/inventory/util/CraftMenus.java

```diff
@@ -34,83 +42,116 @@
-            return CraftMenus.asType(new MenuTypeData<>(EnchantmentView.class, (player, type) -> {
-                return new SimpleMenuProvider((syncId, inventory, human) -> {
-                    return worldAccess(EnchantmentMenu::new).build(player, type);
-                }, Component.empty()).createMenu(player.nextContainerCounter(), player.getInventory(), player);
-            }));
+            return asType(new MenuTypeData<>(EnchantmentView.class, () -> new CraftAccessLocationInventoryViewBuilder<>(handle, EnchantmentMenu::new)));
         }
         if (menuType == MenuType.FURNACE) {
-            return CraftMenus.asType(new MenuTypeData<>(FurnaceView.class, tileEntity(FurnaceBlockEntity::new, Blocks.FURNACE)));
+            return asType(new MenuTypeData<>(FurnaceView.class, () -> new CraftBlockEntityInventoryViewBuilder<>(handle, Blocks.FURNACE, FurnaceBlockEntity::new)));
         }
         if (menuType == MenuType.GRINDSTONE) {
-            return CraftMenus.asType(new MenuTypeData<>(InventoryView.class, worldAccess(GrindstoneMenu::new)));
+            return asType(new MenuTypeData<>(InventoryView.class, () -> new CraftAccessLocationInventoryViewBuilder<>(handle, GrindstoneMenu::new)));
         }
         // We really don't need to be creating a tile entity for hopper but currently InventoryType doesn't have capacity
         // to understand otherwise
         if (menuType == MenuType.HOPPER) {
-            return CraftMenus.asType(new MenuTypeData<>(InventoryView.class, tileEntity(HopperBlockEntity::new, Blocks.HOPPER)));
+            return asType(new MenuTypeData<>(InventoryView.class, () -> new CraftBlockEntityInventoryViewBuilder<>(handle, Blocks.HOPPER, HopperBlockEntity::new)));
         }
         // We also don't need to create a tile entity for lectern, but again InventoryType isn't smart enough to know any better
         if (menuType == MenuType.LECTERN) {
-            return CraftMenus.asType(new MenuTypeData<>(LecternView.class, tileEntity(LecternBlockEntity::new, Blocks.LECTERN)));
+            return asType(new MenuTypeData<>(LecternView.class, () -> new CraftBlockEntityInventoryViewBuilder<>(handle, Blocks.LECTERN, LecternBlockEntity::new)));
         }
         if (menuType == MenuType.LOOM) {
-            return CraftMenus.asType(new MenuTypeData<>(LoomView.class, CraftMenus.STANDARD));
+            return asType(new MenuTypeData<>(LoomView.class, () -> new CraftStandardInventoryViewBuilder<>(handle)));
         }
         if (menuType == MenuType.MERCHANT) {
-            return CraftMenus.asType(new MenuTypeData<>(MerchantView.class, CraftMenus.STANDARD));
+            return asType(new MenuTypeData<>(MerchantView.class, () -> new CraftMerchantInventoryViewBuilder<>(handle)));
+        }
+        if (menuType == MenuType.SHULKER_BOX) {
+            return asType(new MenuTypeData<>(InventoryView.class, () -> new CraftBlockEntityInventoryViewBuilder<>(handle, Blocks.SHULKER_BOX, null)));
```

Comment:
Any reason to not have an underlying Shulkerbox block entity? this means getType on the view doesn't return the right type and putting a shulkerbox inside the container has a weird behavior.

Reply:
It isn't needed in this case as the block entity Is only used for faking it's open. See Anvil impl for example where it's used. In the case of opening a real shulker box the block entity is obtained from the world. And to be completely honest I don't care about InventoryType. This may be controversial but I personally have no plans to build around API this is meant to replace. I'll look into it, but no promises and almost certainly not my priority.

Reply:
We should 100% fix the fact that a shulker menu from an API that is 100% focused around menu types is opened as a generic 9x3 because of CB code relying on the inventory type to figure out the MenuType.

Reply:
Is it opened as a 9x3? I was under the impression only the InventoryType was incorrect. I'll take a look when I get a chance.

Reply:
That method is a real sore spot. In this case I must pass the entity until we go back and replace that internally. Easy fix I guess when I implemented it I ignored it because I didn't care about the distinction with the fake menus. Either way yeah sloppy as hell impl and easy solution.